packages/blade/src/components/Tour/Tour.web.tsx

```diff
@@ -0,0 +1,167 @@
+/* eslint-disable consistent-return */
+/* eslint-disable @typescript-eslint/explicit-function-return-type */
+import React from 'react';
+import { FloatingPortal } from '@floating-ui/react';
+import type { TourMaskRect, TourProps } from './types';
+import { TourContext } from './TourContext';
+import { TourMask } from './TourMask';
+import { TourPopover } from './TourPopover';
+import { useDelayedState } from './utils';
+import { usePrevious, useTheme } from '~utils';
+
+const Tour = ({
+  steps,
+  activeStep,
+  isOpen,
+  onFinish,
+  onOpenChange,
+  onStepChange,
+  children,
+}: TourProps): React.ReactElement => {
+  const { theme } = useTheme();
+  const [refIdMap, setRefIdMap] = React.useState(new Map<string, React.RefObject<HTMLElement>>());
+  const [size, setSize] = React.useState<TourMaskRect>({
+    x: 0,
+    y: 0,
+    height: 0,
+    width: 0,
+  });
+  const transitionDelay = theme.motion.duration.gentle;
+  // delayed state is used to let the transition finish before reacting to the state changes
+  const delayedAciveStep = useDelayedState(activeStep, transitionDelay);
+  const delayedSize = useDelayedState(size, transitionDelay);
+
+  const prevActiveStep = usePrevious(activeStep);
+  const [isTransitioning, setIsTransitioning] = React.useState(false);
+
+  // Keep track of when we are transitioning between steps
+  React.useEffect(() => {
+    if (prevActiveStep === undefined) return;
+    setIsTransitioning(true);
+    const timeout = window.setTimeout(() => {
+      setIsTransitioning(false);
+    }, transitionDelay);
+
+    return () => {
+      window.clearTimeout(timeout);
+    };
+  }, [prevActiveStep, transitionDelay]);
+
+  // main step logic
+  const getCurrentStep = React.useCallback(() => {
+    return steps[activeStep];
+  }, [activeStep, steps]);
+
+  const getCurrentStepRef = React.useCallback(() => {
+    return refIdMap.get(steps[activeStep]?.name);
+  }, [activeStep, refIdMap, steps]);
+
+  const goToNext = React.useCallback(() => {
+    let next = activeStep + 1;
+    if (next >= steps.length) {
+      next = steps.length - 1;
+    }
+    onStepChange?.(next);
+  }, [activeStep, onStepChange, steps.length]);
+
+  const goToPrevious = React.useCallback(() => {
+    let next = activeStep - 1;
+    if (next < 0) {
+      next = 0;
+    }
+    onStepChange?.(next);
+  }, [activeStep, onStepChange]);
+
+  const stopTour = React.useCallback(() => {
+    onFinish?.();
+  }, [onFinish]);
+
+  const attachStep = React.useCallback((id: string, ref: React.RefObject<HTMLElement>) => {
+    if (!ref) return;
+    setRefIdMap((prev) => {
+      return new Map(prev).set(id, ref);
+    });
+  }, []);
```

Comment:
Don't need to pass any dependency here? useCallback will never update the dependent functions/variables inside right?

Reply:
There's no dependent variables inside. 

I'm getting what I need from the useState callback variable: `setRefIdMap((prev) => {})`